### PR TITLE
Implementation of randint and random_integers

### PR DESCRIPTION
**Pull request by delta2323**
The main focus of this PR is the implementation of `cupy.random.randint` and `cup.random.random_integers` and their unit tests. But it also includes the following implementations.
* Fixes of unit tests of `cupy.random.RandomState` (especially those uses mocking objects).
     * Although these tests replaced global functions with mocking objects they left these functions as they are. Because of this, unit tests that follow these tests were forced to use replacing mocking objects instead of original ones. It caused unexpected errors. This PR fixes the problem by restoring original replaced functions.
* Goodness-of-fit test with Pearson's Chi-squared test
* Add more unit tests of `cupy.random.generator` which utilize the goodness-of-fit test module.

### CODE REVIEW

**Comment by delta2323** [comment]
This PR is a feature PR, but changes relating to `FunctionSwitcher` in `RandomState`'s unittestsis a bug fix. So I extract bug fixes and make it new PR as #443

File: cupy/random/generator.py [hunk 1]
```diff
@@ -139,6 +140,60 @@ def random_sample(self, size=None, dtype=float):
...
+            ret &= mask
+            if ret <= mx:
+                break
+        return ret
+
+    def interval(self, mx, size):
+        """Generate multiple integers independently sampled uniformly from [0, mx]
+
+        Args:
+            mx (int): Upper bound of the interval
```
Comment (unnonouno) [inline cupy/random/generator.py]: Insert a period

File: cupy/random/generator.py [hunk 2]
```diff
@@ -139,6 +140,60 @@ def random_sample(self, size=None, dtype=float):
...
+            size (None or int or tuple): Shape of the array or the scalar
+            returned.
+        Returns:
+            int or cupy.array: If ``None``, single integer value is returned.
```
Comment (unnonouno) [inline cupy/random/generator.py]: `cupy.ndarray` ?

File: cupy/random/generator.py [hunk 3]
```diff
@@ -139,6 +140,60 @@ def random_sample(self, size=None, dtype=float):
...
+        if mx == 0:
+            return 0
+
+        mask = mx
+        mask |= mask >> 1
+        mask |= mask >> 2
+        mask |= mask >> 4
+        mask |= mask >> 8
+        mask |= mask >> 16
+        mask |= mask >> 32
```
Comment (unnonouno) [inline cupy/random/generator.py]: How about this code?
```mask = (1 << mx.bit_length()) - 1```

**Comment by unnonouno** [comment]
Do we need to add these functions to the reference document?

File: cupy/random/generator.py [hunk 4]
```diff
@@ -150,6 +150,52 @@ def random_sample(self, size=None, dtype=float):
...
+        ret = cupy.zeros(size, dtype=dtype)
+        sample = cupy.zeros(size, dtype=dtype)
+        done = cupy.zeros(size, dtype=numpy.bool_)
+        while True:
+            curand.generate(
+                self._generator, sample.data.ptr, sample.size)
+            sample &= mask
+            success = sample <= mx
+            ret = ret * (1 - success) + sample * success
+            ret = ret.astype(numpy.int64)
```
Comment (unnonouno) [inline cupy/random/generator.py]: Why is this `astype` required?

File: cupy/random/generator.py [hunk 5]
```diff
@@ -150,6 +150,52 @@ def random_sample(self, size=None, dtype=float):
...
+            mx (int): Upper bound of the interval
+            size (None or int or tuple): Shape of the array or the scalar
+            returned.
+        Returns:
+            int or cupy.ndarray: If ``None``, an ndarray with
+            shape ``()`` is returned.
+            If ``int``, 1-D array of length size is returned.
+            If ``tuple`,` multi-dimensional array with shape
+            ``size`` is returned.
+            Currently, each element of the array is ``numpy.float32``.
```
Comment (unnonouno) [inline cupy/random/generator.py]: ``numpy.int64`` ?

File: cupy/random/generator.py [hunk 6]
```diff
@@ -150,6 +150,47 @@ def random_sample(self, size=None, dtype=float):
...
+            mx (int): Upper bound of the interval
+            size (None or int or tuple): Shape of the array or the scalar
+                returned.
+        Returns:
+            int or cupy.ndarray: If ``None``, an ndarray with
+            shape ``()`` is returned.
+            If ``int``, 1-D array of length size is returned.
+            If ``tuple``, multi-dimensional array with shape
+            ``size`` is returned.
+            Currently, each element of the array is ``numpy.float32``.
```
Comment (unnonouno) [inline cupy/random/generator.py]: ``int64``?

File: cupy/random/generator.py [hunk 7]
```diff
@@ -150,6 +150,47 @@ def random_sample(self, size=None, dtype=float):
...
+        ret = cupy.zeros(size, dtype=dtype)
+        sample = cupy.zeros(size, dtype=dtype)
+        done = cupy.zeros(size, dtype=numpy.bool_)
+        while True:
+            curand.generate(
+                self._generator, sample.data.ptr, sample.size)
+            sample &= mask
+            success = sample <= mx
+            ret = ret * ~success + sample * success
+            ret = ret.astype(numpy.int64)
```
Comment (unnonouno) [inline cupy/random/generator.py]: Why is this `astype` required?

Comment (delta2323) [inline cupy/random/generator.py]: I found that the result of previous line was float. So I forced to convert the type.

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**